operators/pkg/controller/common/settings/canonical_config.go

```diff
@@ -0,0 +1,290 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package settings
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+
+	ucfg "github.com/elastic/go-ucfg"
+	udiff "github.com/elastic/go-ucfg/diff"
+	uyaml "github.com/elastic/go-ucfg/yaml"
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// CanonicalConfig contains configuration for an Elastic resource ("elasticsearch.yml" or "kibana.yml"),
+// as a hierarchical key-value configuration.
+type CanonicalConfig ucfg.Config
+
+// Options are config options for the YAML file. Currently contains only support for dotted keys.
+var Options = []ucfg.Option{ucfg.PathSep(".")}
+
+// NewCanonicalConfig creates a new empty config.
+func NewCanonicalConfig() *CanonicalConfig {
+	return fromConfig(ucfg.New())
+}
+
+// NewCanonicalConfigFrom creates a new config from the API type.
+func NewCanonicalConfigFrom(data untypedDict) (*CanonicalConfig, error) {
+	config, err := ucfg.NewFrom(data, Options...)
+	if err != nil {
+		return nil, err
+	}
+	return fromConfig(config), nil
+}
+
+// MustCanonicalConfig creates a new config and panics on errors.
+// Use for testing only.
+func MustCanonicalConfig(cfg interface{}) *CanonicalConfig {
+	config, err := ucfg.NewFrom(cfg, Options...)
+	if err != nil {
+		panic(err)
+	}
+	return fromConfig(config)
+}
+
+// MustNewSingleValue creates a new config holding a single string value.
+// Convenience constructor, will panic in the unlikely event of errors.
+func MustNewSingleValue(k string, v string) *CanonicalConfig {
+	cfg := fromConfig(ucfg.New())
+	err := cfg.asUCfg().SetString(k, -1, v, Options...)
+	if err != nil {
+		panic(err)
+	}
+	return cfg
+}
+
+// ParseConfig parses the given configuration content into a CanonicalConfig.
+// Expects content to be in YAML format.
+func ParseConfig(yml []byte) (*CanonicalConfig, error) {
+	config, err := uyaml.NewConfig(yml, Options...)
+	if err != nil {
+		return nil, err
+	}
+	return fromConfig(config), nil
+}
+
+// SetStrings sets key to string vals in c.  An error is returned if key is invalid.
+func (c *CanonicalConfig) SetStrings(key string, vals ...string) error {
+	if c == nil {
+		return errors.New("config is nil")
+	}
+	switch len(vals) {
+	case 0:
+		return errors.New("Nothing to set")
+	default:
+		for i, v := range vals {
+			err := c.asUCfg().SetString(key, i, v, Options...)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Unpack returns a typed config given a struct pointer.
+func (c *CanonicalConfig) Unpack(cfg interface{}) error {
+	if reflect.ValueOf(cfg).Kind() != reflect.Ptr {
+		panic("Unpack expects a struct pointer as argument")
+	}
+	return c.asUCfg().Unpack(cfg, Options...)
+}
```

Comment:
I updated the return of `Unpack` to remove the dependency to `estype.ElasticsearchSettings`.
https://github.com/elastic/cloud-on-k8s/blob/f00602603757aed48c15da96c557b40edb9a4dbe/operators/pkg/controller/elasticsearch/settings/canonical_config.go#L91-L95